File: internal/customrules/customrules.go
// Package customrules lets an org define its own detection rules
// declaratively (YAML: resource type + attribute/block + regex pattern),
// without this tool ever executing arbitrary org-supplied code. A DSL
// interpreter with no eval/exec surface is a deliberate security boundary:
// this project runs inside other people's CI pipelines, so "let customers
// write code that runs here" is not a trade-off worth taking for a
// convenience feature.
package customrules

import (
	"fmt"
	"regexp"

	"github.com/foadtalsi/tf-predeploy-firewall/internal/parser"
	"github.com/foadtalsi/tf-predeploy-firewall/internal/report"
	"github.com/foadtalsi/tf-predeploy-firewall/internal/rules"
	"github.com/foadtalsi/tf-predeploy-firewall/internal/schema"
	"gopkg.in/yaml.v3"
)

// Rule is one custom detection rule, as authored in YAML.
type Rule struct {
	ID           string `yaml:"id"`
	ResourceType string `yaml:"resource_type"`       // exact type, or "*" for any resource
	Block        string `yaml:"block,omitempty"`     // nested block type to search within (e.g. "ingress"); omit to check the resource's top-level attributes
	Attribute    string `yaml:"attribute,omitempty"` // attribute name to check; omit to just flag every match of resource_type/block
	Pattern      string `yaml:"pattern,omitempty"`   // regex tested against the attribute's literal value
	Negate       bool   `yaml:"negate,omitempty"`    // flag when the pattern does NOT match, or the attribute is absent — for "must have X" rules
	Severity     string `yaml:"severity"`
	Message      string `yaml:"message"`

	compiled *regexp.Regexp
}

// Config is a full custom rule set, as loaded from an org's YAML config.
type Config struct {
	Rules []Rule `yaml:"custom_rules"`
}

// Load parses and validates a custom rule set. Every rule is validated at
// load time (valid severity, valid regex, required fields present) so a
// typo in an org's config fails the scan loudly instead of silently
// matching nothing.
func Load(data []byte) (*Config, error) {
	var cfg Config
	if err := yaml.Unmarshal(data, &cfg); err != nil {
		return nil, fmt.Errorf("parsing custom rules: %w", err)
	}
	for i := range cfg.Rules {
		if err := cfg.Rules[i].validate(); err != nil {
			return nil, fmt.Errorf("custom rule %d: %w", i, err)
		}
	}
	return &cfg, nil
}

var validSeverities = map[string]bool{"low": true, "medium": true, "high": true, "critical": true}

func (r *Rule) validate() error {
	if r.ID == "" {
		return fmt.Errorf("id is required")
	}
	if r.ResourceType == "" {
		return fmt.Errorf("resource_type is required")
	}
	if !validSeverities[r.Severity] {
		return fmt.Errorf("severity must be one of low/medium/high/critical, got %q", r.Severity)
	}
	if r.Message == "" {
		return fmt.Errorf("message is required")
	}
	if r.Pattern == "" && r.Attribute != "" {
		return fmt.Errorf("attribute %q is set but pattern is empty — a pattern is required to evaluate the attribute's value (omit both to just flag the resource's/block's presence)", r.Attribute)
	}
	if r.Pattern != "" {
		compiled, err := regexp.Compile(r.Pattern)
		if err != nil {
			return fmt.Errorf("invalid pattern: %w", err)
		}
		r.compiled = compiled
	}
	return nil
}

// ruleSet adapts Config into a rules.Rule so it drops straight into the
// same engine (rules.Run) as every built-in rule.
type ruleSet struct{ cfg *Config }

// AsEngineRule wraps a loaded Config as a rules.Rule for rules.Run.
func (c *Config) AsEngineRule() rules.Rule {
	return ruleSet{cfg: c}
}

func (rs ruleSet) Check(in rules.FileInput, aws *schema.AWS) []report.Finding {
	var findings []report.Finding
	for _, res := range in.HeadResources {
		for _, r := range rs.cfg.Rules {
			findings = append(findings, r.check(in.Path, res)...)
		}
	}
	return findings
}

func (r *Rule) check(path string, res *parser.Resource) []report.Finding {
	if r.ResourceType != "*" && r.ResourceType != res.Type {
		return nil
	}

	if r.Block != "" {
		var findings []report.Finding
		for _, b := range res.Blocks {
			if b.Type != r.Block {
				continue
			}
			if f, ok := r.checkAttrs(path, res, b.Attributes, b.Range.Start.Line); ok {
				findings = append(findings, f)
			}
		}
		return findings
	}

	if f, ok := r.checkAttrs(path, res, res.Attributes, res.DefRange.Start.Line); ok {
		return []report.Finding{f}
	}
	return nil
}

// checkAttrs evaluates one rule against one set of attributes (either a
// resource's top-level attributes or one nested block's), returning a
// finding and true if it fires.
func (r *Rule) checkAttrs(path string, res *parser.Resource, attrs map[string]*parser.Attribute, fallbackLine int) (report.Finding, bool) {
	line := fallbackLine
	matched := false

	if r.Attribute == "" {
		// No attribute specified: this rule flags the mere presence of the
		// resource/block (e.g. "don't use aws_iam_user, use aws_iam_role").
		matched = true
	} else {
		attr, present := attrs[r.Attribute]
		switch {
		case present && attr.IsLiteral && r.compiled != nil:
			line = attr.Range.Start.Line
			matched = r.compiled.MatchString(attr.RawValue) != r.Negate
		case present && !r.Negate:
			// Attribute exists but isn't a literal we can pattern-match (a
			// variable/expression) — can't evaluate it, so don't guess.
			matched = false
		case !present:
			// Missing attribute only "matches" for a negated rule (i.e. "this
			// attribute must be present and match the pattern").
			matched = r.Negate
		}
	}

	if !matched {
		return report.Finding{}, false
	}

	return report.Finding{
		File:     path,
		Line:     line,
		Category: report.Category("custom:" + r.ID),
		Severity: report.Severity(r.Severity),
		Resource: res.Address(),
		Message:  r.Message,
	}, true
}

File: src/tfpdf/customrules.py
"""Declarative custom detection rules loaded from local configuration."""

from __future__ import annotations

import re
from dataclasses import dataclass, field

import yaml

from .parser import Attribute, Resource
from .report.finding import Finding, Severity
from .rules import FileInput
from .schema import KnowledgeBase

VALID_SEVERITIES = frozenset({"low", "medium", "high", "critical"})


class CustomRuleError(ValueError):
    """A custom ruleset that could not be parsed or validated."""


@dataclass(slots=True)
class Rule:
    """A custom detection rule as declared in YAML."""

    id: str = ""
    # Exact resource type, or * for every resource.
    resource_type: str = ""
    # Optional nested block type; omit to inspect top-level attributes.
    block: str = ""
    # Optional attribute name; omit to flag the resource or block itself.
    attribute: str = ""
    # Regex tested against the attribute's literal value.
    pattern: str = ""
    # Flag an absent attribute or a nonmatching value for must-have rules.
    negate: bool = False
    severity: str = ""
    message: str = ""

    compiled: re.Pattern[str] | None = field(default=None, repr=False)

    def validate(self) -> None:
        if not self.id:
            raise CustomRuleError("id is required")
        if not self.resource_type:
            raise CustomRuleError("resource_type is required")
        if self.severity not in VALID_SEVERITIES:
            raise CustomRuleError(
                f"severity must be one of low/medium/high/critical, got {self.severity!r}"
            )
        if not self.message:
            raise CustomRuleError("message is required")
        if not self.pattern and self.attribute:
            raise CustomRuleError(
                f"attribute {self.attribute!r} is set but pattern is empty — a pattern is "
                "required to evaluate the attribute's value (omit both to just flag the "
                "resource's/block's presence)"
            )
        if self.pattern:
            try:
                self.compiled = re.compile(self.pattern)
            except re.error as exc:
                raise CustomRuleError(f"invalid pattern: {exc}") from exc

    def check(self, path: str, resource: Resource) -> list[Finding]:
        if self.resource_type != "*" and self.resource_type != resource.type:
            return []

        if self.block:
            findings: list[Finding] = []
            for b in resource.blocks:
                if b.type != self.block:
                    continue
                finding = self._check_attrs(path, resource, b.attributes, b.range.start.line)
                if finding is not None:
                    findings.append(finding)
            return findings

        finding = self._check_attrs(
            path, resource, resource.attributes, resource.def_range.start.line
        )
        return [finding] if finding is not None else []

    def _check_attrs(
        self,
        path: str,
        res: Resource,
        attrs: dict[str, Attribute],
        fallback_line: int,
    ) -> Finding | None:
        """Evaluate a rule against resource attributes or a nested block's attributes."""
        line = fallback_line
        matched = False

        if not self.attribute:
            # Without an attribute filter, flag the resource or block itself.
            matched = True
        else:
            attribute = attrs.get(self.attribute)
            if attribute is not None and attribute.is_literal and self.compiled is not None:
                line = attribute.range.start.line
                matched = bool(self.compiled.search(attribute.raw_value)) != self.negate
            elif attribute is not None and not self.negate:
                # Skip unresolved expressions rather than guessing their values.
                matched = False
            elif attribute is None:
                # An absent attribute matches only a negated must-have rule.
                matched = self.negate

        if not matched:
            return None

        return Finding(
            file=path,
            line=line,
            # Custom categories use custom:<id> strings rather than the closed built-in Category
            # enum.
            category="custom:" + self.id,
            # Prefix custom rule IDs to prevent collisions with built-in IDs.
            rule_name="custom:" + self.id,
            severity=Severity(self.severity),
            resource=res.address(),
            message=self.message,
        )


@dataclass(slots=True)
class Config:
    """A complete custom ruleset loaded from YAML configuration."""

    rules: list[Rule] = field(default_factory=list)

    def check(self, file_input: FileInput, knowledge_base: KnowledgeBase | None) -> list[Finding]:
        """Run custom rules through the same engine interface as built-in rules."""
        findings: list[Finding] = []
        for resource in file_input.head_resources:
            for r in self.rules:
                findings.extend(r.check(file_input.path, resource))
        return findings

    def as_engine_rule(self) -> Config:
        """Return this ruleset as an engine rule; retained for Go API compatibility."""
        return self


def load(data: bytes | str) -> Config:
    """Parse and validate a custom ruleset."""
    try:
        raw = yaml.safe_load(data)
    except yaml.YAMLError as exc:
        raise CustomRuleError(f"parsing custom rules: {exc}") from exc

    if raw is None:
        return Config()
    if not isinstance(raw, dict):
        raise CustomRuleError("custom rules: top level is not a mapping")

    config = Config()
    for i, entry in enumerate(raw.get("custom_rules") or []):
        if not isinstance(entry, dict):
            raise CustomRuleError(f"custom rule {i}: not a mapping")
        rule = Rule(
            id=str(entry.get("id", "")),
            resource_type=str(entry.get("resource_type", "")),
            block=str(entry.get("block", "")),
            attribute=str(entry.get("attribute", "")),
            pattern=str(entry.get("pattern", "")),
            negate=bool(entry.get("negate", False)),
            severity=str(entry.get("severity", "")),
            message=str(entry.get("message", "")),
        )
        try:
            rule.validate()
        except CustomRuleError as exc:
            raise CustomRuleError(f"custom rule {i}: {exc}") from exc
        config.rules.append(rule)
    return config
